### PR TITLE
AgentServer: add invocations_ws (WebSocket) protocol to Invocations

### PR DESCRIPTION
**Pull request by yulin-li**
Adds the `invocations_ws` (WebSocket) transport to
`Azure.AI.AgentServer.Invocations`, so a single `InvocationHandler` can serve
both `POST /invocations` and a full-duplex `/invocations_ws` endpoint on the
same host.

## Why

`Azure.AI.AgentServer.Invocations` previously only served the HTTP
`POST /invocations` transport. Containers that need full-duplex streaming had
no in-process option — every duplex agent had to host its own WebSocket stack
alongside the SDK's HTTP host. This PR closes that gap so multi-protocol
agents share one host, one session, and one process.

### Core (`Azure.AI.AgentServer.Core`, beta.24 Unreleased)
- New `FoundryEnvironment.WebSocketKeepAliveInterval` property bound to the
  `WS_KEEPALIVE_INTERVAL` env var (integer seconds; `Timeout.InfiniteTimeSpan`
  = disabled).
- `AgentHostMiddlewareExtensions.UseAgentServerCore` now also wires
  `IApplicationBuilder.UseWebSockets` with `WebSocketOptions.KeepAliveInterval`
  bound to the env value. Positive values emit RFC 6455 Ping/Pong frames
  (opcode `0x9`) that keep idle WebSocket connections alive across upstream
  proxy / load-balancer idle timeouts.

### Invocations (`Azure.AI.AgentServer.Invocations`, beta.4 Unreleased)
- New virtual `InvocationHandler.HandleWebSocketAsync(WebSocket, InvocationContext, CancellationToken)`.
  Default throws `NotSupportedException`; the endpoint detects non-override
  via cached reflection (`InvocationHandlerCapabilities`) and short-circuits
  to HTTP `404 Not Found` — "endpoint not registered" semantics so a missing
  handler fails fast instead of accepting and immediately closing.
- New `/invocations_ws` route mapped by `MapInvocationsServer`. Lifecycle:
  accept upgrade → invoke handler → close with mapped RFC 6455 code (`1000`
  normal, `1011` internal error). Handler-initiated close codes are preserved
  unchanged.
- Session ID honours `FOUNDRY_AGENT_SESSION_ID` so HTTP and WebSocket
  transports on the same container correlate to the same session, with a
  UUID fallback when the platform doesn't inject one.
- **Structured close-event log** — a single `LogInformation` per connection
  with the documented dotted-name fields
  `azure.ai.agentserver.invocations_ws.{session_id, close_code, duration_ms[, error.code]}`.
  Handler exception text flows through `LogError(ex, ...)` only — never into
  the close-event line — so application stack traces do not leak into the
  structured metric stream. Verified by a unit test that hooks an in-memory
  `ILoggerProvider` and asserts on the captured state-bag keys.
- **No framework-level OpenTelemetry span** for the connection — ASP.NET Core
  auto-propagates the inbound W3C trace context to the request `Activity`,
  so user-created spans inside `HandleWebSocketAsync` are parented correctly
  without a per-connection wrapper. Session / invocation / `x-request-id`
  baggage is added via `InvocationsActivitySource.PropagateInvocationBaggage`,
  the same helper the HTTP `POST /invocations` endpoint uses.
- Tightened cancellation classification: the `OperationCanceledException`
  filter now compares `oce.CancellationToken == httpContext.RequestAborted`
  so a handler-internal OCE (e.g., from its own timeout `CancellationTokenSource`)
  is correctly surfaced as `close_code=1011` instead of being misclassified
  as a clean close during concurrent server shutdown.

### Samples (`Azure.AI.AgentServer.Invocations/samples`)
- `SampleWs1_Echo.md` — same handler serves HTTP `POST /invocations` and
  WebSocket `/invocations_ws` (the simplest multi-protocol agent).
- `SampleWs2_BidirectionalStreaming.md` — full-duplex agent with concurrent
  per-prompt token streamers, mid-flight `cancel`, and graceful `bye`
  shutdown. Demonstrates that the read loop and writer tasks share one
  socket and run independently — the defining property of full-duplex.
- Both samples are compiled-snippet-backed (`tests/Snippets/SampleWs{1,2}Snippets.cs`)
  and exercised by end-to-end tests in `tests/SampleEndToEndTests.cs` per
  AGENTS.md §4.

### Tests (+21 new tests across both packages)
- `Azure.AI.AgentServer.Core/tests/FoundryEnvironmentTests.cs` — 5 tests
  covering `WebSocketKeepAliveInterval` env parsing (unset / positive /
  zero / negative / unparseable).
- `Azure.AI.AgentServer.Invocations/tests/WebSocketEndpointTests.cs` — 11
  tests covering 404 for non-overriding handler, 404 for HTTP GET against
  unregistered route, 400 for plain HTTP GET against WS handler, echo
  round-trip, handler exception → `1011`, handler-initiated close
  preservation, session-ID env precedence + UUID fallback, OCE token-identity
  discrimination, and structured close-event log field-name verification
  (both normal close and exception paths).
- `Azure.AI.AgentServer.Invocations/tests/SampleEndToEndTests.cs` — 5
  end-to-end tests: README WS handler (echo), Sample WS1 HTTP + WS echo,
  Sample WS2 streaming + bye, Sample WS2 mid-flight cancel.
- All new tests use `TaskCompletionSource`-backed gates / `WaitAsync(timeout)`
  for synchronization — no blind `Task.Delay` polling per AGENTS.md §0.IV.

### Docs / API
- `Invocations/README.md` — new `### WebSocket protocol (invocations_ws)`
  section with a compiled snippet and the `WS_KEEPALIVE_INTERVAL` reference
  table.
- `Invocations/CHANGELOG.md` and `Core/CHANGELOG.md` — Unreleased entries.
- API baselines regenerated via `eng/scripts/Export-API.ps1 agentserver`.
- Snippet markdown regenerated via `eng/scripts/Update-Snippets.ps1 agentserver`.

## Verification

| Command | Result |
|---|---|
| `dotnet build Azure.AI.AgentServer.sln` | 0 warnings, 0 errors |
| `dotnet build Azure.AI.AgentServer.sln /p:BuildSnippets=true` | 0 warnings, 0 errors |
| `dotnet test Azure.AI.AgentServer.sln --filter TestCategory!=Live -f net8.0` | **2,235 passed** (175 Core + 97 Invocations + 1963 Responses) |
| `dotnet test Azure.AI.AgentServer.sln --filter TestCategory!=Live -f net10.0` | **2,235 passed** (same) |
| CI (Azure Pipelines `net - pullrequest`) | **21/21 checks green** across Windows / Ubuntu / macOS × net8.0 / net9.0 / net10.0 / net462 matrix |

### Live `websocat` round-trip
Beyond `TestServer`, both samples were driven against a real Kestrel listener
with [`websocat`](https://github.com/vi/websocat):

- **SampleWs1 (echo)** — `printf 'hello\nsecond line\n' | websocat ws://…/invocations_ws`
  echoed each frame back; server log:
  `invocations_ws connection closed: session_id=… close_code=1000 duration_ms=…`.
- **SampleWs2 (bidirectional streaming)** — overlapping `prompt p1` / `prompt p2` +
  `cancel p1` + `bye` produced interleaved `p1`/`p2` token frames on the wire
  (proving full-duplex), a `{"type":"cancelled","id":"p1"}` after the cancel
  took effect, `p2` continuing uninterrupted to its own `done` frame, and
  a clean `close_code=1000` shutdown.

## Compatibility

- Public API surface is **additive only**: `FoundryEnvironment.WebSocketKeepAliveInterval`
  property + `InvocationHandler.HandleWebSocketAsync` virtual method.
- Existing HTTP-only handlers are unaffected: the default `HandleWebSocketAsync`
  throws `NotSupportedException` and the endpoint short-circuits to 404
  before the handler is ever invoked, so no consumer behaviour changes.
- WebSocket Ping/Pong keep-alive is **disabled by default** (preserves the
  spec contract); set `WS_KEEPALIVE_INTERVAL` to a positive integer seconds
  value to enable.

## Prerequisites already in main

This PR depends on two earlier merges and exercises their public surfaces:

| PR | Provides |
|---|---|
| [#59285](https://github.com/Azure/azure-sdk-for-net/pull/59285) | `SseKeepAliveSession` hoisted into `Core`; consumed by `InvocationEndpointHandler` for symmetric HTTP-SSE keep-alive. |
| [#59050](https://github.com/Azure/azure-sdk-for-net/pull/59050) | `InvocationsActivitySource.PropagateInvocationBaggage` (replaced the per-request `invoke_agent` span); the new WS endpoint reuses this helper so HTTP and WS produce identical baggage shapes. |

---

🤖 Authored-by: Copilot CLI (v1.0.49-1)
Model: Claude Opus 4.7 1M (`claude-opus-4.7-1m-internal`)

### CODE REVIEW

File: sdk/agentserver/Azure.AI.AgentServer.Invocations/src/Internal/WebSocketEndpointHandler.cs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+                closeCode = InvocationsWebSocketConstants.CloseInternalError;
+                errorCode = InvocationsWebSocketConstants.ErrorCodeAcceptFailed;
+                throw;
+            }
+
+            try
+            {
+                await handler.HandleWebSocketAsync(webSocket, context, httpContext.RequestAborted);
+            }
+            catch (OperationCanceledException) when (httpContext.RequestAborted.IsCancellationRequested)
```
Comment (yulin-li) [inline sdk/agentserver/Azure.AI.AgentServer.Invocations/src/Internal/WebSocketEndpointHandler.cs]: **Medium**: ambiguous `OperationCanceledException` attribution.

The filter `when (httpContext.RequestAborted.IsCancellationRequested)` only checks **whether** `RequestAborted` is cancelled at the moment of catch — it doesn't verify the OCE was **caused by** `RequestAborted`. If a handler throws an OCE from its own `CancellationTokenSource` (e.g. an internal timeout) at the same moment the server is shutting down, this filter swallows the handler's bug as a clean close (code `1000`) instead of an internal error (`1011`):

```csharp
// Handler:
using var timeoutCts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
await DoWorkAsync(timeoutCts.Token);   // OCE thrown by handler's own timeout
// If server shutdown signals RequestAborted concurrently, the OCE.CancellationToken
// is the handler's token but RequestAborted.IsCancellationRequested is true — wrong close code.
```

The rare-but-real failure mode masks a real handler bug behind a normal close, and the close-event log line will report `close_code=1000` without an `error_code`.

**Suggested fix** — compare the OCE's own token to the one we passed to the handler so only true shutdown OCEs are swallowed:

```csharp
catch (OperationCanceledException oce)
    when (oce.CancellationToken == httpContext.RequestAborted
          && httpContext.RequestAborted.IsCancellationRequested)
```

This is also consistent with the broader .NET idiom for distinguishing cancellation sources (e.g. [HttpClient timeout vs caller cancellation](https://learn.microsoft.com/dotnet/api/system.net.http.httpclient.timeout#remarks)). Worth a small unit test that asserts `1011` is reported when the handler throws OCE from an unrelated token.

_This comment was generated by an AI assistant on behalf of the PR author._

**Review by yulin-li — COMMENTED** [review]
Self-review pass. One medium-severity concern around cancellation-source attribution in `WebSocketEndpointHandler` — left inline. Everything else (resource cleanup, capability detection, reflection AOT, route ordering, FoundryEnvironment parsing, public API surface) checked out.

_Posted by Copilot CLI on behalf of the PR author._

File: sdk/agentserver/Azure.AI.AgentServer.Invocations/src/Internal/WebSocketEndpointHandler.cs (hunk 1 above)
Comment (yulin-li) [reply sdk/agentserver/Azure.AI.AgentServer.Invocations/src/Internal/WebSocketEndpointHandler.cs]: Addressed in commit [`6562f47`](https://github.com/Azure/azure-sdk-for-net/pull/59325/commits/6562f47ea09ee387e716c97301273f7fa8b8a0ff). The filter now compares `oce.CancellationToken == httpContext.RequestAborted` so a handler-internal OCE bound to an unrelated token surfaces as `1011` instead of being silently classified as a clean close. Locked down by the new `HandlerThrowsOceFromUnrelatedToken_ClosesWith1011` test in `WebSocketEndpointTests.cs`.

Linked-token tip: if a handler does `CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ...)` and the OCE carries the *linked* token, the identity check will reject it and the connection closes as `1011`. This is a deliberate trade-off — the alternative (matching any cancelled token) would mask real handler bugs during shutdown. Handlers that want to differentiate cancellation sources should `try`/`catch` their own linked-token OCEs and translate them.

_This comment was generated by an AI assistant on behalf of the PR author._

**Review by yulin-li — COMMENTED** [review]
Second-pass review (post-merge of #59050 + commit `6562f47`).

## Verdict: ✅ clean — no new actionable findings

Checked seven focus areas:

1. **`PropagateInvocationBaggage` integration** — correctly used in WS handler; HTTP and WS paths now produce the same baggage shape; previously missing `x-request-id` propagation is now in.
2. **OCE filter** — token-identity check is correct for the common case. Linked-token edge case is an acceptable trade-off (alternative would mask real handler bugs during shutdown). Locked down by the new `HandlerThrowsOceFromUnrelatedToken_ClosesWith1011` test.
3. **`W3CBaggagePropagator` middleware ordering** — registered before `UseWebSockets()` in `UseAgentServerCore`, so client-sent `baggage` headers are parsed into `Activity.Current` *before* the WS upgrade. User spans inside the handler inherit it correctly.
4. **A365 tracing export** — no problematic interaction; WS doesn't create per-connection spans, user-created spans inside the handler are independent of connection lifetime.
5. **Tests dropped by #59050 merge** — all 4 removed tests asserted on the now-removed `StartInvocationActivity` API. Nothing useful was lost.
6. **Cross-handler consistency** — WS handler intentionally doesn't call `InvocationsExceptionFilter.RecordException(Activity.Current, ex)` because the request Activity may have already closed by the time a long-lived WS connection ends. Logging via `ILogger` is the correct approach for WS.
7. **CHANGELOG bullets** — accurately match the implementation.

## Verification on this checkout (`6562f47`)

| Command | Result |
|---|---|
| `dotnet build Azure.AI.AgentServer.sln` | 0 warnings, 0 errors |
| `dotnet test -f net10.0` | **2,230 passed** (175 Core + 92 Invocations + 1963 Responses) |
| `dotnet test -f net8.0` | **2,230 passed** (same) |

Ready for human review when you are.

_Posted by Copilot CLI on behalf of the PR author._

File: sdk/agentserver/Azure.AI.AgentServer.Invocations/src/Internal/InvocationsWebSocketConstants.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    public const string AttrSpanSessionId = "azure.ai.agentserver.invocations_ws.session_id";
+
+    /// <summary>Structured-log field key carrying the final RFC 6455 close code.</summary>
+    public const string AttrSpanCloseCode = "azure.ai.agentserver.invocations_ws.close_code";
+
+    /// <summary>Structured-log field key carrying the connection duration in milliseconds.</summary>
+    public const string AttrSpanDurationMs = "azure.ai.agentserver.invocations_ws.duration_ms";
+
+    /// <summary>Structured-log field key carrying a short error tag.</summary>
+    public const string AttrSpanErrorCode = "azure.ai.agentserver.invocations_ws.error.code";
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/Azure.AI.AgentServer.Invocations/src/Internal/InvocationsWebSocketConstants.cs:50]: Valid finding — this was a real divergence between the documented cross-SDK wire contract and the actual log shape. Fixed in [`d9914e07`](https://github.com/Azure/azure-sdk-for-net/pull/59325/commits/d9914e077e6) by changing the `LogInformation` templates to use the dotted `AttrSpan*` constants directly as placeholder names. Built as compile-time `const string` concatenation so the template stays a constant (CA2254-safe):

```csharp
private const string CloseEventTemplate =
    "invocations_ws connection closed: session_id={" + InvocationsWebSocketConstants.AttrSpanSessionId +
    "} close_code={" + InvocationsWebSocketConstants.AttrSpanCloseCode +
    "} duration_ms={" + InvocationsWebSocketConstants.AttrSpanDurationMs + "}";
```

Locked down by two new tests in `WebSocketEndpointTests.cs`:

* `CloseEventLog_UsesDottedStructuredFieldNames` — wires a `CapturingLoggerProvider` into the test host, runs a clean WS round-trip, and asserts `closeEvent.StateKeys` contains `azure.ai.agentserver.invocations_ws.{session_id, close_code, duration_ms}`.
* `CloseEventLog_IncludesErrorCodeField_WhenHandlerThrows` — same setup with a throwing handler, asserts `error.code` field is present with value `internal_error`, and verifies the rendered message does **not** contain the handler exception text (`"boom"`), confirming the close-event log does not leak application stack traces.

The captured-log helper uses a `TaskCompletionSource`-backed `WaitForMatchAsync(timeout)` instead of `Task.Delay`-polling per AGENTS.md §0.IV.

_This comment was generated by an AI assistant on behalf of the PR author._

File: sdk/agentserver/Azure.AI.AgentServer.Invocations/tests/WebSocketEndpointTests.cs [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+            // object, so we don't assert on them directly here. Driver-level checks
+            // live in higher-level integration scenarios. The smoke test is that
+            // the upgrade still succeeds when we set the header from the handler.
+            Assert.That(ws.State, Is.EqualTo(WebSocketState.Open));
+        }
+        finally
+        {
+            await app.StopAsync();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (yulin-li) [reply sdk/agentserver/Azure.AI.AgentServer.Invocations/tests/WebSocketEndpointTests.cs:422]: Valid — the test name promised coverage of the upgrade-response header but only asserted `ws.State == WebSocketState.Open` (its own comment admitted it didn't inspect upgrade headers). Dropped in [`d9914e07`](https://github.com/Azure/azure-sdk-for-net/pull/59325/commits/d9914e077e6).

The session-ID-on-upgrade-response code path is exercised live by the `websocat` round-trip documented in an earlier commit (`SessionId` env var feeding through `WebSocketEndpointHandler.HandleAsync` lines 83–88 → response headers), and the related handler-visible session-ID precedence is covered by the existing `SessionId_HonoursFoundryAgentSessionIdEnv` and `SessionId_FallsBackToGeneratedUuid_WhenEnvUnset` tests.

_This comment was generated by an AI assistant on behalf of the PR author._

**Review by niuzheng168 — APPROVED** [review]

File: sdk/agentserver/Azure.AI.AgentServer.Core/CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Release History
 
-## 1.0.0-beta.24 (2026-05-15)
+## 1.0.0-beta.24 (Unreleased)
```
Comment (ankitbko) [inline sdk/agentserver/Azure.AI.AgentServer.Core/CHANGELOG.md]: Change it to today's date so we can publish?

File: sdk/agentserver/Azure.AI.AgentServer.Invocations/CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Release History
 
-## 1.0.0-beta.4 (2026-05-15)
+## 1.0.0-beta.4 (Unreleased)
```
Comment (ankitbko) [inline sdk/agentserver/Azure.AI.AgentServer.Invocations/CHANGELOG.md]: same?

**Review by ankitbko — CHANGES_REQUESTED** [review]

File: sdk/agentserver/Azure.AI.AgentServer.Core/CHANGELOG.md (hunk 4 above)
Comment (yulin-li) [reply sdk/agentserver/Azure.AI.AgentServer.Core/CHANGELOG.md]: Done in [`52d89702`](https://github.com/Azure/azure-sdk-for-net/pull/59325/commits/52d897021a0) — `Azure.AI.AgentServer.Core 1.0.0-beta.24` heading now reads `(2026-05-19)`.

_This comment was generated by an AI assistant on behalf of the PR author._

File: sdk/agentserver/Azure.AI.AgentServer.Invocations/CHANGELOG.md (hunk 5 above)
Comment (yulin-li) [reply sdk/agentserver/Azure.AI.AgentServer.Invocations/CHANGELOG.md]: Done in [`52d89702`](https://github.com/Azure/azure-sdk-for-net/pull/59325/commits/52d897021a0) — `Azure.AI.AgentServer.Invocations 1.0.0-beta.4` heading now reads `(2026-05-19)`.

_This comment was generated by an AI assistant on behalf of the PR author._

**Review by ankitbko — APPROVED** [review]
